extensions-api/src/main/scala/pl/touk/nussknacker/engine/api/expression/ExpressionParseError.scala

```diff
@@ -0,0 +1,108 @@
+package pl.touk.nussknacker.engine.api.expression
+
+import org.springframework.expression.spel.SpelNode
+import pl.touk.nussknacker.engine.api.typed.typing.{TypedDict, TypedNull, TypingResult, Unknown}
+
+trait ExpressionParseError {
+  def message: String
+}
+
+object ExpressionParseError {
+  sealed trait ArgumentTypeError extends ExpressionParseError {
+    val found: List[TypingResult]
+    val functionName: String
+
+    def expectedString: String
+
+    protected def typesToString(types: List[TypingResult]): String =
+      types.map(_.display).mkString(", ")
+
+    override def message: String =
+      s"Mismatch parameter types. Found $functionName(${typesToString(found)}). Required: $functionName($expectedString)"
+  }
+
+  case class NoVarArgumentTypeError(expected: List[TypingResult],
+                                    found: List[TypingResult],
+                                    functionName: String) extends ArgumentTypeError {
+    override def expectedString: String = typesToString(expected)
+  }
+
+  case class VarArgumentTypeError(expected: List[TypingResult],
+                                  expectedVarArgument: TypingResult,
+                                  found: List[TypingResult],
+                                  functionName: String) extends ArgumentTypeError {
+    override def expectedString: String = typesToString(expected :+ expectedVarArgument) + "..."
+  }
+
+  case class ExpressionTypeError(expected: TypingResult, found: TypingResult) extends ExpressionParseError {
+    override def message: String = s"Invalid type of expression: expected ${expected.display}, found ${found.display}"
+  }
+
+  object InvocationOnUnknownError extends ExpressionParseError {
+    override def message: String = s"Method invocation on ${Unknown.display} is not allowed"
+  }
+
+  object InvocationOnNullError extends ExpressionParseError {
+    override def message: String = s"Method invocation on ${TypedNull.display} is not allowed"
+  }
+
+  case class UnknownMethodError(methodName: String, displayableType: String) extends ExpressionParseError {
+    override def message: String = s"No method named $methodName in type $displayableType"
+  }
+
+  case class UnknownClassError(className: String) extends ExpressionParseError {
+    override def message: String = s"Class $className does not exist"
+  }
+
+  case class InvalidMethodReference(methodName: String) extends ExpressionParseError {
+    override def message: String = s"Invalid method reference $methodName"
+  }
+
+  case class TypeReferenceError(refClass: String) extends ExpressionParseError {
+    override def message: String = s"$refClass is not allowed to be passed as TypeReference"
+  }
+
+  case class IllegalProjectionSelectionError(typ: TypingResult) extends ExpressionParseError {
```

Comment:
this one (and probably a few others?) are spel-specific. They shouldn't  be part of extensions-api